tools/run_tests/python_utils/jobset.py

```diff
@@ -474,11 +473,10 @@ def finish(self):
     while self._running:
       if self.cancelled(): pass  # poll cancellation
       self.reap()
-    # Clear the alarms when finished to avoid a race condition causing job
-    # failures. Don't do this when running multi-VM tests because clearing
-    # the alarms causes the test to stall
-    if platform_string() != 'windows' and self._clear_alarms:
-      signal.alarm(0)
+    global have_alarm
```

Comment:
Ran into the missing timeouts when testing #12210 and can confirm this PR fixes.

One nit: since `signal.alarm()` cancels and overwrites any pre-existing alarms, can we completely get rid of the global `have_alarm` variable and `def alarm_handler`? I think cancellation could use `signal.alarm(0)` then

Reply:
I'm not sure how getting rid of `have_alarm` will allow us to use `signal.alarm(0)` again. To be fair though, I'm still not sure why `signal.alarm(0)` is causing issues, but getting rid of it has gotten tests to time out properly.

Reply:
IIC, currently `alarm(0)` is cancelling the alarm but not setting `have_alarm = False`, and the following jobset then has no pending alarm but thinks it does. But I think we can unconditionally set alarm in `reap` and cancel it in `finish`?

Reply:
I added a commit to test your change. I think it should work. This code was added years ago by @ctiller. Craig, do you happen to remember why you needed to have a global variable to track whether an alarm was active? I think alarms should be unique per `jobset.py` processes.